src/java/org/apache/cassandra/locator/InetAddressAndPort.java

```diff
@@ -382,4 +385,83 @@ public long serializedSize(InetAddressAndPort from, int version)
             }
         }
     }
+
+    // Serializer for handling FWD_FRM message parameters. Pre-4.0 deserialization is a special
+    // case in the message
+    public static final class FwdFrmSerializer implements IVersionedSerializer<InetAddressAndPort>
+    {
+        public static final FwdFrmSerializer fwdFrmSerializer = new FwdFrmSerializer();
+        private FwdFrmSerializer() { }
+
+        public void serialize(InetAddressAndPort endpoint, DataOutputPlus out, int version) throws IOException
+        {
+            byte[] buf = endpoint.addressBytes;
+
+            if (version >= MessagingService.VERSION_40)
+            {
+                out.writeByte(buf.length + 2);
+                out.write(buf);
+                out.writeShort(endpoint.port);
+            }
+            else
+            {
+                out.write(buf);
+            }
+        }
+
+        public long serializedSize(InetAddressAndPort from, int version)
+        {
+            //4.0 includes a port number
+            if (version >= MessagingService.VERSION_40)
+            {
+                if (from.address instanceof Inet4Address)
+                    return 1 + 4 + 2;
+                assert from.address instanceof Inet6Address;
+                return 1 + 16 + 2;
+            }
+            else
+            {
+                if (from.address instanceof Inet4Address)
+                    return 4;
+                assert from.address instanceof Inet6Address;
+                return 16;
+            }
+        }
```

Comment:
@jonmeredith There are certainly chunks of `serialize()` and `serializedSize()` that are duplicated. (i.e. Only their pre-4.0 behavior diverges?) It may be nice to push some of this into a base class w/ abstract `pre40Serialize()` and `pre40SerializedSize()` methods to cut down on surface area until it all goes away if we ever break support for mixed 3.x <-> x.x version clusters.

Reply:
I had the patch that way originally but backed away as each method has to handle pre/post 4.0 messages so to when adding protections to make sure the wrong serializer wasn't called call paths longer/uglier so I went back to the simpler option of having the two full implementations.